src/CoreWCF.Http/src/CoreWCF/Channels/HttpChannelHelpers.cs

```diff
@@ -121,80 +123,62 @@ private void ThrowMaxReceivedMessageSizeExceeded()
 
-        private async Task<Message> ReadBufferedMessageAsync(Stream inputStream)
+        private async Task<Message> ReadMessageAsync(PipeReader reader)
         {
-            ArraySegment<byte> messageBuffer = GetMessageBuffer();
-            byte[] buffer = messageBuffer.Array;
-            int offset = 0;
-            int count = messageBuffer.Count;
-
-            while (count > 0)
+            ReadOnlySequence<byte> buffer = ReadOnlySequence<byte>.Empty;
+            Message message = null;
+            try
             {
-                int bytesRead = await inputStream.ReadAsync(buffer, offset, count);
-                if (bytesRead == 0) // EOF
+                while (true)
                 {
-                    if (ContentLength != -1)
+                    ReadResult result = await reader.ReadAsync();
+                    buffer = result.Buffer;
+
+                    if (buffer.Length >= _settings.MaxReceivedMessageSize)
                     {
-                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                            new ProtocolException(SR.HttpContentLengthIncorrect));
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(MaxMessageSizeStream.CreateMaxReceivedMessageSizeExceededException(_settings.MaxBufferSize));
                     }
 
-                    break;
-                }
-                count -= bytesRead;
-                offset += bytesRead;
-            }
+                    if (result.IsCompleted)
+                    {
+                        break;
+                    }
 
-            return await DecodeBufferedMessageAsync(new ArraySegment<byte>(buffer, 0, offset), inputStream);
-        }
+                    reader.AdvanceTo(buffer.GetPosition(0), buffer.End);
+                }
 
-        private async Task<Message> ReadChunkedBufferedMessageAsync(Stream inputStream)
-        {
-            try
-            {
-                return _messageEncoder.ReadMessage(await BufferMessageStreamAsync(inputStream, _bufferManager, _settings.MaxBufferSize), _bufferManager, ContentType);
+                message = await DecodeBufferedMessageAsync(buffer);
             }
-            catch (XmlException xmlException)
+            finally
             {
-                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                    new ProtocolException(SR.MessageXmlProtocolError, xmlException));
+                reader.AdvanceTo(buffer.Start, buffer.End);
             }
+
+            return message;
         }
 
-        private async Task<Message> ReadStreamedMessageAsync(Stream inputStream)
+        private async Task<Message> ReadStreamedMessageAsync(PipeReader reader)
         {
-            MaxMessageSizeStream maxMessageSizeStream = new MaxMessageSizeStream(inputStream, _settings.MaxReceivedMessageSize);
+            MaxMessageSizeStream maxMessageSizeStream = new(reader.AsStream(), _settings.MaxReceivedMessageSize);
 
```

Comment:
Not wanting to feature creep here so this can happen as a follow up issue as we need to get code completed and merged at some point, but we could create a MaxMessageSizePipeReader which wraps the PipeReader. It would be a small performance improvement, but cleaner.